---
title: "Pruning & Rooting Phylogenetic Trees with ggTree"
author: "Mar Schmidt"
date: "`r Sys.Date()`"
output:
  html_document: 
    code_folding: show
    theme: spacelab
    highlight: pygments
    keep_md: no
    toc: yes
    toc_float:
      collapsed: no
      smooth_scroll: yes
      toc_depth: 3
  keep_md: true  
editor_options: 
  chunk_output_type: console
---

# Goals 

1. Load the [FastTree2](https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0009490) unrooted tree that we calculated in `analysis/04A_Phylogenetic_Tree_Construction.Rmd` and `raw_preprocessed_physeq``.  
2. Merge the unrooted tree to the `raw_preprocessed_phyloseq` object.  
3. Visualize and inspect tree with [ggtree](https://besjournals.onlinelibrary.wiley.com/doi/full/10.1111/2041-210X.12628).
4. Evaluate long branches & prune ASVs, if needed. *If your dataset has long branches, you will need to do this before rooting!!*
5. After pruning, root the tree, ideally within the Archaea and, if not, mid-root the tree.
6. Combine the new, rooted tree with the phyloseq object. 
7. Save 2 phyloseq objects: 1. Unrooted tree phyloseq object, 2. Rooted tree phyloseq object. 

## Inputs 

1. Our previously pre-processed phyloseq object that has chloroplasts, mitochondria, reverse complements, and negative controls removed, which is located in `data/03_PreProcessing/raw_preprocessed_physeq`.
2. The raw, unrooted maximum liklihood tree that we generated using  [FastTree2](https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0009490), which is located in `data/04_PhylogeneticTree/ASVs_unrooted.tree`

## Outputs 

1. The output of this file will be a S4 phloseq objected called `phytree_preprocessed_physeq`, which we will save to `data/04_PhylogeneticTree/`. 

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE,
                      fig.align = "center",
                      # Always relevant to the document directory 
                      # send any figure output to this folder 
                      fig.path = "../figures/04_PhylogeneticTree/",
                      warning = FALSE) 
```

# Before you start

## Timing of this script

Let's record how long this file took to run on the class server, which we will record at the end of the script. 

```{r rmd-start}
# What time did we start running this script? 
start_time <- Sys.time()
```

## Set my seed 
```{r set-seed}
# Any number can be chosen 
set.seed(238428)
```

## Load Packages 
```{r load-packages}
pacman::p_load(tidyverse, phyloseq, ggtree, phytools, tidytree, phylobase,
               install = FALSE)
```

## 1. Load Data files 
```{r load-data}
# Preprocessed phyloseq object 
load("data/03_PreProcessing/raw_preprocessed_physeq.RData")

# Inspect the phyloseq object 
raw_preprocessed_physeq

# Load in the tree! 
unrooted_tree <- read.tree("data/04_PhylogeneticTree/ASVs_unrooted.tree")

# Take a look at the tree
unrooted_tree
str(unrooted_tree)
```

# 2. Merge Unrooted Tree & `raw_preprocessed_physeq`

Now that we've loaded in both our `raw_preprocessed_physeq` and the unrooted tree, let's merge them into one phyloseq object that can be used to plot the tree with ggtree. 


```{r merge-physeq}
# Intuition check 
stopifnot(ntaxa(raw_preprocessed_physeq) == ntaxa(unrooted_tree))
# No news is good news! 

# Merge the tree with the phyloseq object 
unrooted_physeq <- 
  merge_phyloseq(raw_preprocessed_physeq, unrooted_tree)

# Let's take a look! 
unrooted_physeq
```


# 3. Visualize and inspect tree with [ggtree](https://besjournals.onlinelibrary.wiley.com/doi/full/10.1111/2041-210X.12628).

Now, let's take a look 

## Kingdom Level Tree

```{r plot-tree-unrooted, fig.width=4.5, fig.height=7}
# Make a basic tree with the domains 
kingdom_tree <- 
  ggtree(unrooted_physeq) + 
  # color tips by kingdom
  geom_tippoint(mapping = aes(color = Kingdom)) + 
  # add a title
  labs(title = "Unrooted Tree") + 
  scale_color_manual(values = c("firebrick3", "cornflowerblue", "grey")) + 
  # Move the legend to the bottom of the tree 
  theme(legend.position = "bottom")

# Look at it 
kingdom_tree

# Check if tree is rooted
is.rooted(unrooted_tree) # should be TRUE
```

In the above unrooted tree, we can see all of the Archaea fall within the bacteria. Huh?! This **does not make evolutionary sense!** Therefore, we will need to **root** the tree to fix this. *But what does it mean to root the tree, you ask???* We will get to this in a minute! 

### Do you expect to have Archeaea? 

Regions of the 16S rRNA gene are better/worse for amplifying archaea that might be present in our samples. If you're wondering more, please check out this [2021 Annual Reviews of Microbiology paper by Tahon et al](https://www.annualreviews.org/content/journals/10.1146/annurev-micro-040921-050212), specifically Figure 2 and the tables. What I've learned about archaea over the years is: 

- Better regions for Archaeal Detection: 
    - V4 or V4–V5 (e.g., 515F-Y / 806R-B or 926R): Improved archaeal coverage (including Thaumarchaeota and Euryarchaeota) compared to older V4-only approaches due to primer updates by [Parada et al., 2015](https://journals.asm.org/doi/10.1128/msystems.00009-15) and [Aprill et al., 2016](https://www.int-res.com/abstracts/ame/v75/n2/p129-137/), captures both Bacteria and Archaea reasonably well and provides the best balance for broad community surveys
- Caution with these regions
    - [Original V4 primers (515F/805R)](https://www.pnas.org/doi/10.1073/pnas.1000080107?url_ver=Z39.88-2003&rfr_id=ori%3Arid%3Acrossref.org&rfr_dat=cr_pub++0pubmed): Common and well-supported for Bacteria BUT original 515F underrepresents Thaumarchaeota, a key archaeal group in soils, oceans, and freshwater. 
    - V3–V4 (e.g., 341F / 805R): Popular due to high taxonomic resolution for Bacteria but biased against Archaea in most environmental samples—specifically, some archaeal groups lack primer binding sites in this region. 
    - V6–V8 or V7–V9: Rarely used for Archaea (less common for bacteria, too), region is less conserved across archaeal groups, Low coverage unless paired with archaeal-specific primers.

In summary: 

### Summary: Hypervariable Region Selection for Archaea

| Region     | Coverage of Archaea | Notes                                            |
|------------|---------------------|--------------------------------------------------|
| V4–V5      | ✅ Good              | Best for mixed Bacteria + Archaea               |
| V3–V5      | ✅ Good              | Broad, longer amplicon (good with 2×300 bp)     |
| V4 only    | ⚠️ Moderate          | Use [515F-Y](https://journals.asm.org/doi/10.1128/msystems.00009-15) instead of original 515F         |
| V3–V4      | ❌ Poor              | Biased toward Bacteria                          |
| V6–V8      | ❌ Poor              | Not recommended for general use                 |

*Ok, now that we've considered the archaea, let's think about the branches in our tree!* 

# 4. Long Branches? 

Now, let's think about the quality of our tree. Evaluate that kingdom tree above...

**Do you have any long branches in your tree?** If so, you will need to prune your ASVs. This becomes especially important if your dataset does not have Archaea (*e.g.,* looking at you 341F V3 primers!) especially if you do not have Archaea and will need to 

<span style="color: red;">INTERPRETATION #1: Are there any suspect branches in the raw, unrooted tree? Or does your tree look “clean”? Why or why not?</span>

Before we move onto rooting, this is where we need to assess if we have any long and wonky branches! 


## 4a. Evaluate Long Branches

This part will get a little manual, however, this is an important step along the way to ensure that we are feeling certain that our data quality is good. 

### Kingdom Tree with Node Labels 

Let's put labels on our tree so we have a bit more flexibility

```{r kingdom-node-tree, fig.width=4.5, fig.height=7}
kingdom_node_tree <- 
  kingdom_tree + 
  # Add the node label 
  geom_text(aes(label=node), hjust= -0.5, vjust = -0.3, size = 2)

# Plot the tree
kingdom_node_tree
```

In the phylogeny above, it looks like **node #703** is quite long (as long as the Archaea!) and towards the top, there's another **node #1910**, which also looks long! Otherwise, things look mostly ok. 


### Evaluate node #703

```{r evaluate-long-branches-703}
# View a specific clade
# Zoom in on node that looks potentially suspicious
viewClade(kingdom_node_tree + 
          labs(title = "Kingdom Tree: Node 703"), 
          node = 703)

# This plot is too zoomed in! Let's try to zoom out by looking at ancestors
# Lets see its ancestors, which will tell us each node as it steps up. 
# It's ideal to look a couple nodes up but not too far!
tidytree::ancestor(unrooted_tree, 703) # The input is the TREE, not phyloseq!!

# Let's evaluate 3308!!
viewClade(kingdom_node_tree, 3308) # even more zoomed out
```

In the subtree above, it's clear that 703 and to a lesser extent 702 are long branches. Let's inspect them some more! 


```{r pull-node703}
# Or if there are two nodes that we can see we can find its MRCA
## create tibble of tree to identify offspring
tree_tib <- as_tibble(unrooted_physeq@phy_tree)

# Inspect
head(tree_tib)
str(tree_tib)

# lets look at the long branch
mrca_node_703 <- tidytree::MRCA(unrooted_tree, .node1 = 703, .node2 = 702) #3315

# Inspect
mrca_node_703

## create dataframe to view ASVs and taxa info and to pull the ASVs
node_3315 <- 
  offspring(tree_tib, mrca_node_703, tiponly = TRUE) %>% 
  as.data.frame() %>% 
  pull(label)

# Inspect
node_3315

# Now, let's pull out the taxonomy information, including the ASV sequence!
node_3315_df <- 
  unrooted_physeq %>%
  tax_table() %>%
  as.data.frame() %>% 
  filter(ASV %in% node_3315) 

# Take a look: 
glimpse(node_3315_df)

# look at counts of the ASVs
unrooted_physeq %>%
  subset_taxa(., ASV %in% node_3315_df$ASV) %>%
  otu_table() %>%
  data.frame() %>%
  colSums()

# Lets BLAST our sequences to investigate if they are real or not
node_3315_df 
```

Interesting! We can already conclude a few things: 

- There are VERY FEW counts! Only a total of 6 counts across three samples. A very rare microbe!
- Only bacterial-level taxonomy assigned to these two ASVs. 

Both of these conclusions make us skeptical of these two ASVs! Currently, I'm not sure if they should be kept in the dataset. Therefore, let's go to blast to help us decide what to do!

# BLAST

Now, copy and paste the ASV sequences into [NCBI's Standard Nucleotide BLAST](https://blast.ncbi.nlm.nih.gov/Blast.cgi?PROGRAM=blastn&BLAST_SPEC=GeoBlast&PAGE_TYPE=BlastSearch). 

## What to look for with BLAST Results? 

If you're identifying a sequence (like 16S rRNA or a gene):

- % Identity ≥ 97% for species-level 16S match (lower for genus)
- E-value ≤ 1e-5 (lower is better—e.g. 1e-20, 1e-50, etc.)
- High query coverage (ideally >90%)
- Long alignment length, especially for full genes or 16S sequences

### Most Important Values in BLAST Output

| Field            | What It Means                                               | Why It Matters                                       | Good Values                         |
|------------------|--------------------------------------------------------------|------------------------------------------------------|-------------------------------------|
| **% Identity**   | % of exact matches in aligned region                         | Measures sequence similarity                         | ≥ 97% for species (16S); lower for genus |
| **E-value**      | Expected number of hits this good by chance                  | Lower = more statistically significant match         | ≤ 1e-5 (e.g., 1e-20, 1e-50, etc.)   |
| **Bit Score**    | Normalized alignment score                                  | Higher = better alignment                            | > 200 for strong alignments         |
| **Alignment Length** | Number of bases or amino acids in the aligned region     | Longer alignments are generally more reliable        | As long as possible (e.g., >250 bp for 16S) |
| **Query Coverage** | % of your sequence aligned with the hit                   | High coverage increases confidence in identification | > 90% (ideally near 100%)           |
| **Accession & Description** | ID and description of the matched sequence       | Tells you what the hit is and where it comes from    | Use top hit + trusted source        |

#### The BLAST results: 

- `ASV_1917`: The first sequence that pops up is an uncultured bacterium from a sample that was from river biofilms in a project titled "Spatial patterns in biofilm diversity across hierarchical levels of over-floodplain landscapes". This seems legit! 
    - **E-value:** 2e-43, which is a very strong alignment! 
    - **% Identity:** 80.95%
- `ASV_2618`: The first sequence that is identified is from a project entitled: "Millions of Boreal Shield Lakes can be used to Probe Archaean Ocean Biogeochemistry". 
    - **E-value:** 2e-23, which means there's a 1 in 5e+22 chance (basically zero) that you would see a match this good by random chance in the database. This indicates a very strong alignment—its trustworthy! Thus, the sequences are almost certainly related (homologous) and not just similar by accident.
    - **% Identity:** 76.33%
    
We see that the two ASVs that are in our sample. ASV_1917 is node 702 and ASV_2618 is the longer node that is node 703. 

**In conclusion...** Our tree looks legitimate and we do not need to remove ASVs that were found in node 3316. BLASTing the two ASVs found that node 703 is long but it is likely a *real* ASV. Thus we do not need to remove anything! :) This is some good diligent science! We should be proud of ourselves. 

### Evaluate node #1910 

```{r evaluate-long-branches-1910}
# View a specific clade
# Zoom in on node that looks potentially suspicious
viewClade(kingdom_node_tree + 
          labs(title = "Kingdom Tree: Node 1910"), 
          node = 1910)

# This plot is too zoomed in! Let's try to zoom out by looking at ancestors
# Lets see its ancestors, which will tell us each node as it steps up. 
# It's ideal to look a couple nodes up but not too far!
tidytree::ancestor(unrooted_tree, 1910) # The input is the TREE, not phyloseq!!

# Let's evaluate 4519!!
viewClade(kingdom_node_tree, 4519) # even more zoomed out
```

In the sub-tree above, it's clear that 1910 is longer than the nearby branches. Let's inspect some more! 

```{r pull-node1910}
# lets look at the long branch
mrca_node_1910 <- tidytree::MRCA(unrooted_tree, .node1 = 1910, .node2 = 1909) #3315

# Inspect
mrca_node_1910 #4529

## create dataframe to view ASVs and taxa info and to pull the ASVs
node_4529 <- 
  offspring(tree_tib, mrca_node_1910, tiponly = TRUE) %>% 
  as.data.frame() %>% 
  pull(label)

# Inspect
node_4529

# Now, let's pull out the taxonomy information, including the ASV sequence!
node_4529_df <- 
  unrooted_physeq %>%
  tax_table() %>%
  as.data.frame() %>% 
  filter(ASV %in% node_4529) 

# Take a look: 
glimpse(node_4529_df)

# look at counts of the ASVs
unrooted_physeq %>%
  subset_taxa(., ASV %in% node_4529_df$ASV) %>%
  otu_table() %>%
  data.frame() %>%
  colSums()

# Lets BLAST our sequences to investigate if they are real or not
node_4529_df 
```

Interesting! We can already conclude a few things: 

- There are VERY FEW counts of these two ASVs! Only a total of 5 counts across two samples, likely found only in the one sample. This helps further show how zero-inflated these microbial datasets are. 
- In terms of taxonomy, these ASVs are assigned to the Order level: a Burkholderiales and a Chlamydiales. 

If we want, we can also confirm by BLASTing: 

- `ASV_2350`: Had an e-value of 5e-93 and a percent identity of 92.65%, source matches the sample type here
- `ASV_2286`: Has an e-value of 3e-71 and a percent identity of 87.35%, source matches the sample type for this study, too

We can keep these in for now. :) 

## 4b. Prune your tree

Note that our tree actually looks quite nice and we do not need to prune any ASVs out. 

However, if you find in your own data set that you have ASVs that need to be removed here is the code for how to do that!

```{r prune-ASVs, eval = FALSE}
# Function from Joey McMurdie: https://github.com/joey711/phyloseq/issues/652
pop_taxa = function(physeq, badTaxa){
  allTaxa <-  taxa_names(physeq)
  allTaxa <- allTaxa[!(allTaxa %in% badTaxa)]
  return(prune_taxa(allTaxa, physeq))}

# Let's use the pop_taxa function :) 
# Recreate a phyloseq object without your unwanted ASV or node
# Lets use the example ASVs above from node_3315_df
unrooted_physeq_rm3315 <- 
  unrooted_physeq %>%
  pop_taxa(., node_3315_df$ASV)

# Check it 
unrooted_physeq_rm3315

# Intuition Check: How many differ???
ntaxa(unrooted_physeq) - ntaxa(unrooted_physeq_rm3315) 

# Visually inspect 
ggtree(unrooted_physeq_rm3315) + 
  geom_tippoint(mapping = aes(color = Kingdom))
```

*Note:* this is just an example of how to remove two ASVs because we actually don't need to prune this tree! :) This is likely because we don't have reverse complements and also because we were very stringent with our ASV lengths! 


<span style="color: red;">INTERPRETATION #2: Were there any long branches in the unrooted tree? If so, did you decide to remove any ASVs from your tree? Why or why not? What data convinced you to remove each of the ASVs?</span>

# 5. Root the Tree

Now that we've pruned the trees of potentially contaminating ASVs, it's time to root! 

*What does it mean to root a tree, you ask?* **Rooting a tree** means designating a specific node in a phylogenetic tree as the common ancestor of all other nodes. This establishes a direction of evolution — from the root (the oldest ancestor) to the tips (the current taxa or sequences). A tree without a root is called **unrooted**, and it only shows relationships (like how closely related things are), not the direction of descent.

## Why do we root trees in microbial community analyses?

In microbial ecology, rooted trees are especially useful for:

1. **Calculating Phylogenetic Diversity Metrics**: Metrics like Faith’s Phylogenetic Diversity and the UniFrac distances require a rooted tree to calculate how much evolutionary history is shared between communities. Without a root, the distances between taxa don’t make biological sense in terms of lineage or shared ancestry.
2. **Comparing Community Structure:** Rooting allows you to identify lineages that dominate or are lost between samples, trace evolutionary shifts in microbial composition, interpret ecological or functional patterns across a gradient (*e.g.,* salinity, pH, antibiotic exposure). 
3. **Biological Interpretation** Rooted trees are required for the following interpretations, which microbial community ecologists often want to know: 
    - *Are these microbes derived from a common ancestor?*
    - *Do the samples diverge from a known lineage (e.g., Proteobacteria)*?
    - *What’s the evolutionary placement of unknown organisms?*

## Methods of Rooting Trees 

So, **how do we root a tree**? There are two different types of methods to root a tree: 

1. **Outgroup rooting:** Choose a known lineage outside your group of interest (*e.g.,* Archaea as outgroup to Bacteria). ***This is always preferred when we have a known outgroup!***
    - *Pros:*
        - Biologically meaningful — places the root based on known evolutionary relationships.
        - More accurate if your outgroup is correctly chosen and aligns well.
        - **Best for:** Hypothesis-driven studies, evolutionary interpretations, or when your dataset includes taxa from multiple domains/phyla.
    - *Cons:* 
        - Requires a reliable outgroup — too distant, and it can distort the tree (long-branch attraction).
        - Not always available or appropriate (e.g., if you're only analyzing a subset of taxa without known outgroups).

2. **Midpoint rooting:** Places the root at the midpoint of the longest path in the tree (useful when no clear outgroup exists). 
    - *Pros:*
        - Simple and automatic — no need for prior biological knowledge.
        - Works well for diverse microbial datasets where outgroup info is unclear or unavailable.
        - **Best for**: Exploratory analyses, diversity metrics (like UniFrac), or when working with just bacteria/just archaea and you don’t have a good outgroup.
    - *Cons:*
        - Assumes a molecular clock — that all lineages evolve at similar rates.
        - Can misplace the root if there’s rate variation across branches.

**In Summary:**

| Rooting Method      | Best When...                                    | Pros                        | Cons                                 |
|---------------------|--------------------------------------------------|-----------------------------|--------------------------------------|
| **Outgroup**         | You have a well-placed, closely related outgroup | Biologically accurate       | Needs prior knowledge; may bias tree |
| **Midpoint**         | No good outgroup; exploratory or metric-focused | Fast and automatic          | Assumes equal evolutionary rates     |

**In summary:** If you're working in QIIME or another microbiome pipeline and don't have a clear outgroup, **midpoint rooting** is typically fine — especially for things like UniFrac or beta diversity. If you’re doing evolutionary or taxonomic analysis and **have a suitable outgroup, use that instead**.


## 5a. Archaeal Root

As discussed above, if you have archaea in your sample, it's best to use the outgroup method to root the tree. Since we have archaea in this dataset, we will use the outgroup method. However, below we will walk through an example of how to mid-root the tree. (*Note that midpoint rooting is not preferable when there's a clear outgroup!*)

```{r archaeal-root}
# Is the tree rooted? 
tidytree::is.rooted(phy_tree(unrooted_physeq))

# but first find the ASVs that are Archaea
archaeal_ASVs <- 
  tax_table(unrooted_physeq) %>%
  as.data.frame() %>%
  filter(Kingdom == "Archaea") %>%
  pull(ASV)

# How many archaeal ASVs? 
length(archaeal_ASVs)
archaeal_ASVs

# find the node that encompasses all Archaea
archaeal_node <- 
  phytools::findMRCA(unrooted_tree, 
                     tips = archaeal_ASVs, type = "node") 

# Inspect
archaeal_node # node = 3394

# View clade to check it out
viewClade(kingdom_tree, archaeal_node)
viewClade(kingdom_node_tree, archaeal_node)
```

```{r archeal-node-highligh, fig.width=4.5, fig.height=7}
# Let's highlight! 
node_3394 <- 
  ggtree(unrooted_physeq) + 
  geom_tippoint(mapping = aes(color = Kingdom), size = 1.5) + 
  #geom_text(aes(label = node), hjust = -0.7, vjust = -0.2, size = 2) +
  geom_highlight(node = archaeal_node, fill = "goldenrod1") + 
  scale_color_manual(values = c("firebrick3", "cornflowerblue", "grey"))

# Visualize it 
node_3394

# Root it! 
archaeal_rooted_tree <-
  ape::root(phy_tree(unrooted_physeq), 
            node = archaeal_node, resolve.root = TRUE)

# Check it! 
tidytree::is.rooted(archaeal_rooted_tree)
```

Great, we have successfully rooted the tree! Now, let's plot it to also confirm this visually. 

### 6a. Merge phyloseq: Archeal Rooted Tree

```{r merge-archeal-root-physeq}
# Merge with physeq 
archaeal_rooted_physeq <- 
  merge_phyloseq(raw_preprocessed_physeq, archaeal_rooted_tree)
```


### Plot Archaeal Rooted Tree 

```{r plot-archaeal-rooted-tree, fig.width=4.5, fig.height=7}
# Finally let's plot it! 
# Plot rooted simple tree with no text 
archaeal_rooted_tree_plot <- 
  ggtree(archaeal_rooted_physeq) + 
  geom_tippoint(aes(color = Kingdom)) +   
  # add a title
  labs(title = "Archaeal Rooted Tree") + 
  scale_color_manual(values = c("firebrick3", "cornflowerblue", "grey"))

# View it! 
archaeal_rooted_tree_plot

# Plot rooted tree with text
archaeal_rooted_node_tree_plot <- 
  archaeal_rooted_tree_plot +
  geom_text(aes(label = node), hjust = -0.5, vjust = -0.3, size = 1.5) 

# View it! 
archaeal_rooted_node_tree_plot
```

In the tree above, we can still see the longer branches in the dataset but we can feel good because we inspected them! :) 

<span style="color: red;">INTERPRETATION #3 (If archaeal rooting): Now that you've rooted your tree, does the tree seem "finalized"? Are there any other branches/ASVs to inspect? Why or why not?</span>


## 5b. Midpoint Root 

*If you have Archaea or a clear outgroup, always use the outgroup rooting. If not, you can midpoint root.*

As mentioned above, if your project's data does not have Archaea or a specified outgroup, it is possible to midpoint root the tree. Midpoint rooting calculates tip to tip distances and then places the root halfway between the two longest tips. This is why it is important to **first prune your tree of contaminating lineages!** If contaminating lineages are included in the midrooting of the tree, then it will throw off the branch lengths. 

**Before midrooting the tree:**

1. Investigate and pull the ASVs in your long branches and BLAST to confirm if these are real bacterial ASVs, as mentioned above in the "long branches" section.
2. Remove any unwanted ASVs using the prune function. **Long nodes (branches) are outliers**! If we keep them in the tree, bias will be introduced into the tree reconstruction creating the **long branch attraction** error where distantly related taxa appear closely related due to a high rate of evolution on long branches. 

If you **do not** have Archaea and **have** removed your long branches, then you may follow the code below to midroot your tree.

### 6b. Merge phyloseq: Midpoint Rooted Tree

```{r midpoint-root-tree-physeq}
# Check to see if tree is rooted.. expect it to be false
is.rooted(phy_tree(unrooted_physeq))

# Let's midpoint root the tree
midpoint_rooted_tree <- 
  unrooted_physeq %>%
  phy_tree() %>%
  phytools::midpoint.root()

# Is the new tree rooted? Should be TRUE!
is.rooted(midpoint_rooted_tree)

# Assign to a new phyloseq object!
# Merge tree with the raw_preprocessed physeq
midpoint_rooted_physeq <-
  merge_phyloseq(raw_preprocessed_physeq, midpoint_rooted_tree)
```

Now, we have a merged phyloseq object! Let's take a look at what it looks like and see how it compares with the archaeal rooted tree above. 

### Plot Midpoint Rooted Tree

```{r plot-midpoint-root-tree,fig.width=4.5, fig.height=7}
# Quick inspection of tree 
midpoint_rooted_tree_plot <-
  ggtree(midpoint_rooted_physeq) + 
  geom_tippoint(mapping = aes(color = Kingdom)) + 
  geom_tippoint(aes(color = Kingdom)) + 
  labs(title = "Midpoint Rooted Tree") + 
  scale_color_manual(values = c("firebrick3", "cornflowerblue", "grey"))

# Look at the tree 
midpoint_rooted_tree_plot

# Add nodes 
midpoint_rooted_tree_node_plot <- 
  midpoint_rooted_tree_plot + 
  geom_text(aes(label = node), hjust = -0.5, vjust = -0.3, size = 1.5) 

# View it 
midpoint_rooted_tree_node_plot
```

You'll see that even with a midpoint root, it's still rooted within the Archaea. This is because the Archaea were the longest branches in the unrooted tree. However, you'll notice that the branch lengths are quite different from the Archaeal rooted tree. This is because the Archeal rooted tree assumes ancestry, which is not an assumption in midrooting. 

<span style="color: red;">INTERPRETATION #3 (If midpoint rooting): Now that you've rooted your tree, does the tree seem "finalized"? Are there any other branches/ASVs to inspect? Why or why not?</span>

## 7. Save phyloseq objects

The phyloseq objects can be saved in one file as a list of objects. Or, we can save them individually. Two examples are below: 

### Save multiple phyloseq objects as a list

This will keep two phyloseq objects in on .RData object that will allow us to easily load in both trees into an R session at once. 

```{r save-phyloseq-objects-list}
# Take a quick look at the two!
archaeal_rooted_physeq # Archaeal rooted tree
midpoint_rooted_physeq # Midpiont rooted tree

# Save both phyloseq objects to one .RData file 
save(list = c("archaeal_rooted_physeq", "midpoint_rooted_physeq"),
     file = "data/04_PhylogeneticTree/phytree_preprocessed_physeqs.RData")
```

### Save individual phyloseq objects

Here, we will create two individual files that will each have a single phyloseq object. This is useful if we only want to work with one tree. (For example, we will move forward with the archaeal tree.) 

```{r save-phyloseq-object}
# Now, save archaeal rooted phyloseq object
save(archaeal_rooted_physeq, 
     file = "data/04_PhylogeneticTree/archaeal_rooted_physeq.RData")

# Now, save midpoint rooted phyloseq object
save(midpoint_rooted_physeq, 
     file = "data/04_PhylogeneticTree/midpoint_rooted_physeq.RData")
```

Now that we have our rooted trees inside of a phyloseq object, we can go forth and begin our next analysis!

# Final info for Reproducibility 

## Check Render Time
```{r stop-time}
# Take the time now that we are at the end of the script
end_time <- Sys.time()
end_time 

# Echo the elapsed time
elapsed_time <- round((end_time - start_time), 3)
elapsed_time
```

## Session Information

```{r session-info}
# Ensure reproducibility with package version information
devtools::session_info()
```
